### ISSUE TITLE
State of Torch Named Tensors

### ISSUE THREAD

**Issue by hjalmarlucius** [pytorch/pytorch#60832]
What is the development status and plans on Named Tensors? I don't see any news or progress on it for the last six months at least. I want to use this feature for a system I'm building and am considering whether to build on top of Alex Rush' lib (http://nlp.seas.harvard.edu/NamedTensor) or extending the native but experimental feature. I would certainly prefer the latter but its codebase is C and I'm unsure of how robust or supported it is / will become.

**Comment by zou3519** [comment]
@hjalmarlucius we're not actively developing the feature due to a lack of bandwidth for resolving the ambiguities in the design. We understand that there is still a good deal of user interest in the feature and would appreciate any feedback and help. Could you tell us a bit more about how you wanted to use named tensors in your system?

**Comment by hjalmarlucius** [comment]
Thanks for your response.
* What I'm building: A low-code system for reinforcement learning, primarily for industrial applications.
* Why I want named dimensions:
  * Reduce complexity: Rearrangement operations (transpose etc) become implicit in the `tensor1.dot(tensor2, dimnames)`. Less documentation for me to write and less stuff for users to learn.
  * Fewer false positives: No accidentally permitted operations because shapes are broadcastable but undesired.
  * Integration: When plotting / showing outputs, dimension names are nearly always required and now they are a freebie.

As for the ambiguities in the design, three inputs:
* In the long term, I'm convinced that named dimensions are better in every way. Particularly outside of research where people of different backgrounds are to understand what happens. If you want to grow pytorch outside of research, I would integrate named dimensions into pytorch itself.
* I think named and unnamed syntax is quite different. They can be made compatible by taking the union of all operations but I think they are better implemented as two strict domains with a shared backend. Therefore, if you don't want to go fully named in the future, I would rather opt for named dimensions as a separate wrapper akin to Alex Rush's lib.
* If going with an external wrapper, I think xarray has a great framework to integrate against. See this thread for reference: https://github.com/pydata/xarray/issues/3232

**Comment by Pedrexus** [comment]
Any development in this regard or plans to make it a stable feature?

**Comment by albertz** [comment]
We have something similar (but actually far more powerful) as named tensors in our framework [RETURNN](https://github.com/rwth-i6/returnn) (a generic deep learning framework, originally mostly for speech recognition and handwriting recognition, then extended for machine translation and language modeling, but basically anything now). The RETURNN implementation and our experience might be helpful to resolve any ambiguities. We called them dimension tags. Dimension tags also evolved in RETURNN over many years, and along the way, we also had to resolve a couple of ambiguities, but now I think most of it is resolved.

Dimension tags are more than just names. It's an object instead of a string, and this object can contain other information. The most important use case is a time axis, which contains sequence lengths per batch index, so sequence lengths are of shape [B], but in general, we allow sequence lengths again of any shape, and their shape is again described via dim tags.

A bit on the history in RETURNN:

We had some Theano backend initially (since 2013). Then we started the TensorFlow backend in 2016. When we started the TF backend, we introduced the `Data` class, which is basically a wrapper around a `tf.Tensor` but containing other meta information on the tensor, such as `batch_dim_axis`, `time_dim_axis`, `feature_dim_axis`, potentially sequence lengths (assumed of shape [B]) for the potential dynamic time dim axis, and some other information, such as a potential vocab of number of classes for class indices. This was already very flexible and allowed us many efficient combinations of LSTMs with other type of models, and each layer would automatically permute the tensor to be in its most efficient format. No layer did assume some specific input shape. Instead, every layer just assumed some dimensions to be present in the input (e.g. B, T, F) but in any order.

Then in 2017, when we did our first attention-based encoder-decoder implementation, a tensor could have multiple dynamic sequence lengths, and that is when we introduced dim tags, but initially only for those dynamic dims.

Then some time later we consistently used dim tags for all dimensions.

As a consequence, most layers operating on some specific dimension/axis (i.e. a lot of layers, `LinearLayer`, `DotLayer`, `ConvLayer`, `SelfAttentionLayer`, `ReduceLayer`, ...) now could refer to the axis via dimension tag. This assumes that the dim tag is unique in a tensor, which is almost always the case. We have some operations on the dim tag itself, such as creating a new copy of it, which you want for self-attention, to be able to distinguish the two dims.

Remember that we still don't have any requirement on the order of dimensions. This is such that every layer can just permute the dims in a way that is most efficient for its operation, and then it would leave it that way. E.g. the LSTM returns [T,B,F], no matter the input. But it does not really matter, as no layer depends on the order of dimensions.

Consistent use of dim tags, and arbitrary order of dimensions, this means that for all non-elemwise operations, i.e. layers/functions operating on one or multiple dimensions, you always need to specify on what dimensions you want to operate. This makes the code quite a bit more verbose than what you usually see in other frameworks, but it really avoids a bunch of problems, and it allows for a number of optimizations. As an example, [here is a Conformer implementation](https://github.com/rwth-i6/returnn_common/blob/04f24f42c63e2ef4a73a80c709276a4da0a1cd10/nn/conformer.py#L362), where you see the use of those dim tags (just `nn.Dim` there).

**Comment by albertz** [comment]
What ambiguities do you actually have?

One problem currently is that it's the named tensor support in PyTorch is very incomplete, for many primitive ops you get errors, and thus it's basically not usable (you don't want to put `tensor.rename_(None)` everywhere just to work around errors you otherwise get).

**Comment by hjalmarlucius** [comment]
What you're describing summarizes the issue well, it's just not usable in its current form and sometimes also raises errors.

**Comment by Jack-Yu-815** [comment]
I'd love to see Named Tensor get support from primitive ops in PyTorch and become usable. The benefit of Named Tensor is unequivocal. Developers are already naming their tensors by commenting on their shapes anyway. Using Named Tensor 
  1. adds more semantics to a tensor because tensor names are symbols whereas tensor shapes are numbers (more intuitive shape inspection)
  2. makes tensor operations safer (due to name-checking that happens before) and easier to identify shape-related errors
  3. will likely gain immediate developer adoption because developers already have a habit of naming tensors by adding comments (but it's weak and not enforced unlike Named Tensor)

As a grad student and AI researcher, I feel Named Tensor is a game changer for all developers who work with tensors. I'd love to provide more feedback and make contributions to Named Tensor development. Anyone can point me to the right direction?

What are the ambiguities in design as @zou3519 mentioned?

**Comment by hjalmarlucius** [comment]
@Jack-Yu-815 your points are sound and since 2021 I've written an extensive library for this since I think it's better off as a standalone downstream lib ala Keras than an integrated component of pytorch. It's not public atm as it's part of a much bigger codebase for my RL/simulation startup but I plan to open source it eventually.

As for the ambiguities I can confirm there are many - from top of mind:
* first and foremost: combining unnamed and named tensors is very prone to errors. named tensors's dim order could[should] be an implementation detail but they are a first order citizen in unnamed tensors.
* elementwise ops are fine and most reduce+transform ops are mostly just a `dimnum = self.names.index(dimname)` away
* index ops are quite different (and IMO much better) with names (you could e.g. envisage neat indexing with `dict[str, NamedLongTensor]` as a gather-like op)
* autograd ops like jacobians should be changed to accommodate that you can specify which dimnames to batch the diff over
* I'd argue the more complex an operator is, the more different it becomes. With dimnames you don't need to assume similar shapes so you have more room for variably-shaped input.
* a very different ambiguity that still bothers me: Should you permit non-unique dimnames on B to allow C to preserve A's names in`C=A@B`? But that opens a can of worms everywhere else.

As a fourth benfit on on your list (or extending point 2), I'd add typability. If you ignore dim order, many shape-changing operations are fully typable shapes using today's typing spec (though with quite a few overloads needed to handle scalars and dtype trees), see the below screenshot:
![image](https://github.com/pytorch/pytorch/assets/35001974/22febc41-94ad-4532-8dda-c1f535378207)

**Comment by albertz** [comment]
For reference, there is also [`torchdim`](https://github.com/facebookresearch/torchdim) / [`functorch.dim`](https://github.com/pytorch/pytorch/tree/main/functorch/dim), which looks to me like a better alternative than the current named tensors in PyTorch. Here, `Dim` is an object, not just a string. (Quite similar as `Dim` is an object in RETURNN.)

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/160236
